crates/artifact_content/src/file_system_state.rs

```diff
@@ -0,0 +1,530 @@
+use std::{
+    collections::{HashMap, HashSet},
+    path::PathBuf,
+};
+
+use crate::operation_text::hash;
+use common_lang_types::{
+    ArtifactFileName, ArtifactHash, ArtifactPathAndContent, FileContent, FileSystemOperation,
+    SelectableName, ServerObjectEntityName,
+};
+use isograph_config::PersistedDocumentsHashAlgorithm;
+
+#[derive(Debug, Clone, Default)]
+pub struct FileSystemState {
+    root_files: HashMap<ArtifactFileName, (FileContent, ArtifactHash)>,
+    nested_files: HashMap<
+        ServerObjectEntityName,
+        HashMap<SelectableName, HashMap<ArtifactFileName, (FileContent, ArtifactHash)>>,
+    >,
+}
+
+impl FileSystemState {
+    pub fn insert(&mut self, path_and_content: &ArtifactPathAndContent) {
+        let value = (
+            FileContent::from(path_and_content.file_content.clone()),
+            ArtifactHash::from(hash(
+                &path_and_content.file_content,
+                PersistedDocumentsHashAlgorithm::Md5,
+            )),
+        );
+
+        match &path_and_content.artifact_path.type_and_field {
+            Some(type_and_field) => {
+                self.nested_files
+                    .entry(type_and_field.parent_object_entity_name)
+                    .or_default()
+                    .entry(type_and_field.selectable_name)
+                    .or_default()
+                    .insert(path_and_content.artifact_path.file_name, value);
+            }
+            None => {
+                self.root_files
+                    .insert(path_and_content.artifact_path.file_name, value);
+            }
+        }
+    }
+
+    pub fn from_artifacts(artifacts: Vec<ArtifactPathAndContent>) -> Self {
```

Comment:
there's no need to make these functions methods on `FileSystemState`, IMO. It's generally cleaner to have pure functions.

In the `from_artifacts` case, it might be cleanest to have `impl From<Vec<ArtifactPathAndContent>> for FileSystemState`. But IMO `insert` can just be a free fn (if needed, it also doesn't seem too complicated to just inline into `impl From<Vec<...>>`